BattleCity/sources/protocols/fieldmessage.cpp

```diff
@@ -0,0 +1,447 @@
+#include "protocols\fieldmessage.h"
+
+#include <sstream>
+#include <cassert>
+#include <algorithm>
+#include <bitset>
+#include <limits>
+using namespace std;
+
+#include <iostream>
+
+namespace pv {
+
+	#define INT_FIELD &int_type_name_
+	#define STRING_FIELD &string_type_name_
+
+	constexpr size_t MESSAGE_SIZE_BYTES = 8;
+	constexpr size_t MASK_BYTES = 8;
+	constexpr size_t FIELD_TYPE_BYTES = 1;
+	constexpr size_t STRING_LEN_BYTES = 2;
+	constexpr size_t INT_VALUE_BYTES = 4;
+																						  //================value============/		
+	constexpr int MIN_LEN_MESS_BYTES = static_cast<int>(MESSAGE_SIZE_BYTES + MASK_BYTES + (FIELD_TYPE_BYTES + INT_VALUE_BYTES) + FIELD_TYPE_BYTES + STRING_LEN_BYTES);
+	constexpr int NEW_INT_FIELD_BYTES = static_cast<int>(FIELD_TYPE_BYTES + INT_VALUE_BYTES);
+	
+	constexpr int NEW_STR_FIELD_BYTES = static_cast<int>(FIELD_TYPE_BYTES + STRING_LEN_BYTES);
+
+	const char INT_FIELD_BYTECODE = 0;
+	const char STR_FIELD_BYTECODE = 127;
+
+	FieldMessage::FieldMessage()
+		: field_to_type_p_{ 
+			{Fields::MESSAGE_SIZE, INT_FIELD},
+			{Fields::CORR_ID, STRING_FIELD},
+			{Fields::LIGHT_TANK_HEALTH, INT_FIELD},
+			{Fields::CURRENT_PLAYER_NAME, STRING_FIELD},
+			{Fields::GAME_OVER_STR_RESOURCE, STRING_FIELD},
+			{Fields::HIGHEST_SCORE, INT_FIELD},
+			{Fields::__LIMIT_FIELD__, INT_FIELD}
+		},
+		  fields_{
+			  { Fields::MESSAGE_SIZE, MIN_LEN_MESS_BYTES },
+			  { Fields::CORR_ID, "FieldMessage#" + std::to_string(message_id_)}
+		},
+		  enums_masks_(MAX_FIELDS_SUPPORTED_COUNT)
+		{
+			assert(sizeof(int) == INT_VALUE_BYTES);
+			for (size_t i = 0; i < MAX_FIELDS_SUPPORTED_COUNT; ++i) {
+				FieldsUnderlyingType enumValue = GET_ENUMERATOR_VALUE(i);
+				enums_masks_[i] = enumValue;
+			}
+			++message_id_;
+
+			get<int>(fields_[Fields::MESSAGE_SIZE]) += get<string>(fields_[Fields::CORR_ID]).size(); 
+		}
+
+	bool FieldMessage::FieldHasType(Fields field, const std::string* typeP) const {
+		auto mapIterator = field_to_type_p_.find(field);
+		if (mapIterator == field_to_type_p_.end()) {
```

Comment:
You can simplify that looks like:
```suggestion
		if (auto mapIterator = field_to_type_p_.find(field); mapIterator == field_to_type_p_.end()) {
```

Reply:
Or less idiomatic 
```cxx
if (auto mapIterator = field_to_type_p_.find(field) != field_to_type_p_.end())
```